### PR TITLE
fix/feat: use joins instead of n+1/opening multiple connections

### PR DESCRIPTION
**Pull request by markphelps**
## Overview

This is kind of a doozy so bear with me. 🐻

This came about while I was looking at our StackHawk scans in reference to FLI-177.

Basically, StackHawk issues a bunch of concurrent queries to probe our app for security concerns, SQL injections, etc.

When I was running this locally (using the SQLite DB for Flipt), I kept running into these 500 errors:

```console
ERROR	finished unary call with code Internal	{"server": "grpc", "grpc.start_time": "2023-01-27T15:40:34-05:00", "system": "grpc", "span.kind": "server", "grpc.service": "flipt.Flipt", "grpc.method": "GetRule", "peer.address": "127.0.0.1:63802", "error": "rpc error: code = Internal desc = database is locked", "grpc.code": "Internal", "grpc.time_ms": 5191.929}
```

```
 database is locked
```

Being the relevant bit.

## A trip to the DB

In looking into this further and doing some googling, I realized that this is mainly a problem with SQLite and opening multiple connections simultaneously, as SQLite is really just a file, so it makes sense that only 1 connection should hold the lock on the file while doing its querying.

This also led me to https://github.com/mattn/go-sqlite3/issues/274, where they basically say the same thing, and recommend:

1. Always calling `row.Close()` to return the connection back to the pool
2. Setting `db.SetMaxOpenConnections(1)` to enforce this constraint of 1 connection

You can see the deadlock/timeout happening now in [the unit tests](https://github.com/flipt-io/flipt/actions/runs/4032805607/jobs/6932918486#step:4:128) since I followed this advice to set max open connections to `1`:

```
 ?   	go.flipt.io/flipt/internal/storage/oplock/testing	[no test files]
coverage: 67.0% of statements
panic: test timed out after 10m0s
```

## Going through our storage code

This led me to look at our storage (SQL) layer, to ensure we were doing (1), which we are, but then I realized there are several cases where we are trying to open issue multiple queries at once ourselves. Mainly when we try to populate a 1-many relationship such as `flag HAS MANY variants` like [here](https://github.com/flipt-io/flipt/blob/9286fff4b6745e83792be9ac2827292010d665fa/internal/storage/sql/common/flag.go#L44-L73) and [here](https://github.com/flipt-io/flipt/blob/9286fff4b6745e83792be9ac2827292010d665fa/internal/storage/sql/common/flag.go#L87-L146) (GetFlag and ListFlags respectively).

Here and in several other places, we are issuing 1 query to get the parent (flag), then calling this method ([variants](https://github.com/flipt-io/flipt/blob/9286fff4b6745e83792be9ac2827292010d665fa/internal/storage/sql/common/flag.go#L355)) to issue a query to get all the children (variants) for that flag.

The key part is that we don't actually close the flag `Row` before issuing the second query to get all the variants, so we are basically creating a deadlock in the case of SQLite.

While we could work around this for `GetFlag`, by first getting the flag row, then closing the row, then issuing the query to get all variants, this approach won't work as easily for `ListFlags` as we are required to loop through each row to hydrate the flags.

### 1. Joins The Old Fashioned Way

The correct way to do this kind of thing in SQL is to just use JOINs and get all the flags and their variants at once. This has the benefit of not locking the db in our case and also results in fewer queries overall.

This is what I have prototyped in this PR, starting with the `ListFlags` because it is the most 'difficult'.

Unfortunately, Go does not make it easy to work with JOINS using the stdlib or even using `masterminds/squirrel` which is the SQLbuilder library we are using here. This is why I had to add the `map` in this PR to check to see if we have 'seen' this flag before adding it to the results.

Proposal 1 is basically to continue this pattern throughout the entire storage layer for all parent->child relationships, so that we can do things the 'right way' and stop creating unnecessary queries and potential performance degradations/errors in highly concurrent scenarios.

### 2. Introduce a Data Model layer + use SQLx or SQLC

This option is basically introducing data models that map to our database instead of passing the proto models all the way down + introducing a library to help with this parent <-> child relationship, removing the need to do this mapping ourselves.

#### SQLx

[SQLx ](https://github.com/jmoiron/sqlx)supports parent-child relationships in their [StructScan](https://jmoiron.github.io/sqlx/#advancedScanning) helper, it would just require us to define each of these data structs and add the appropriate `db:` struct tags to each field.

The benefits of this approach seem to me to be:

1. Get us closer to where we want to be in FLI-40 (using proper domain models)
2. Prevent the work required to manually handle these joins/relationships
4. Potentially reduce bugs by depending on `StructScan` instead of scanning the row results into the data types ourselves

#### SQLC

SQLC is similar to the approach that SQLx would result in, however, the main difference is that it generates the data models and the query/mutation methods themselves! Ex: https://docs.sqlc.dev/en/latest/howto/select.html

The benefits of this approach seem to me to be:

1. All of the benefits in the SQLx approach above
2. Less code to write/maintain ourselves

The downside of SQLC however is that it does not have the same database compatibility that the Go SQL package/SQLx has. 

Currently, it only supports:

- MySQL
- Postgres
- SQLite (beta)

While these are the only DBs we need now, it could potentially limit us if we wanted to support another SQL db down the road (although I'm not sure if we'll ever want to actually do this).

### Tl;dr

- We should probably use `JOIN`s instead of multiple queries to populate parent child data
- There are a few options, so we should figure out which one we actually want to move to

### CODE REVIEW

File: internal/storage/sql/common/flag.go [hunk 1]
```diff
@@ -118,37 +130,92 @@ func (s *Store) ListFlags(ctx context.Context, opts ...storage.QueryOption) (sto
...
-		flag.CreatedAt = createdAt.Timestamp
-		flag.UpdatedAt = updatedAt.Timestamp
+		f.CreatedAt = fCreatedAt.Timestamp
+		f.UpdatedAt = fUpdatedAt.Timestamp
 
-		if err := s.variants(ctx, flag); err != nil {
-			return results, err
+		// append flag to output results if we haven't seen it yet
+		if _, ok := uniqueFlags[f.Key]; !ok {
+			flags = append(flags, f)
```
Comment (GeorgeMac) [inline internal/storage/sql/common/flag.go:168]: I believe you could actually do all of this without the map. Given we ensure the return rows are ordered by Flag. This might be the natural order of the join. However, we could ensure it in the `ORDER BY` clause. e.g. `ORDER BY f.created_at, f.key DESC`.

Once we have that invariant, we can simply do the following two steps:

1. Compare the newly scanned flag key to the last seen flag key in the result so far.
2. If the flag keys differ, then append the newly scanned flag to the results.
3. `append` the currently scanned variant to the last flag in the results seen so far.

**Comment by GeorgeMac** [comment]
There is another approach to consider. We're doing the classic `N+1` but we could also do the Rails style `preload`.

Rails usually lets you pick between JOIN (`eager_load`) and the two queries style (`preload`).
Or `includes` which decides for you based on some heuristic I have forgotten.

The `preload` method just makes two queries.
The first is for the parent and the second uses `WHERE fk IN (id1, id2, ...)` to load the relation.

**Comment by markphelps** [comment]
Update: we chose to go with `option 1. Joins The Old Fashioned Way` and just continue doing the queries/results scanning by hand.

I continued the approach to the ListSegments and ListRules storage methods.

We are also now able to compare benchmarks added in #1307 (now on main) to those from this PR to see how using JOINS instead of multiple queries affects peformance/memory allocations.

Here is the results:

### Main Branch (no JOINS)

```
goos: linux
goarch: amd64
pkg: go.flipt.io/flipt/internal/storage/sql
cpu: Intel(R) Xeon(R) Platinum 8370C CPU @ 2.80GHz
BenchmarkListFlags/no-pagination-2         	      20	 289475937 ns/op	19884141 B/op	  677228 allocs/op
BenchmarkListFlags/pagination-limit-10-2   	    1681	   3557599 ns/op	  228084 B/op	    7661 allocs/op
BenchmarkListFlags/pagination-limit-25-2   	     763	   7836253 ns/op	  526125 B/op	   17818 allocs/op
BenchmarkListFlags/pagination-limit-100-2  	     204	  28974698 ns/op	 2016770 B/op	   68599 allocs/op
BenchmarkListFlags/pagination-limit-500-2  	      40	 145750363 ns/op	 9963442 B/op	  339415 allocs/op
BenchmarkListFlags/pagination-2            	      40	 147484753 ns/op	 9963565 B/op	  339424 allocs/op
BenchmarkListSegments/no-pagination-2      	       3	1891547016 ns/op	19284261 B/op	  697224 allocs/op
BenchmarkListSegments/pagination-limit-10-2         	     261	  23129548 ns/op	  221543 B/op	    7881 allocs/op
BenchmarkListSegments/pagination-limit-25-2         	     100	  53584245 ns/op	  510681 B/op	   18338 allocs/op
BenchmarkListSegments/pagination-limit-100-2        	      27	 199298428 ns/op	 1956430 B/op	   70620 allocs/op
BenchmarkListSegments/pagination-limit-500-2        	       6	 973445186 ns/op	 9662901 B/op	  349432 allocs/op
BenchmarkListSegments/pagination-2                  	       6	 974635037 ns/op	 9663716 B/op	  349444 allocs/op
```

## This Branch (Joins)

```
goos: linux
goarch: amd64
pkg: go.flipt.io/flipt/internal/storage/sql
cpu: Intel(R) Xeon(R) Platinum 8370C CPU @ 2.80GHz
BenchmarkListFlags/no-pagination-2         	      51	 115100076 ns/op	17045878 B/op	  695298 allocs/op
BenchmarkListFlags/pagination-limit-10-2   	    2778	   2183352 ns/op	   29532 B/op	    1031 allocs/op
BenchmarkListFlags/pagination-limit-25-2   	    2511	   2382021 ns/op	   54766 B/op	    2075 allocs/op
BenchmarkListFlags/pagination-limit-100-2  	    1878	   3163501 ns/op	  181777 B/op	    7290 allocs/op
BenchmarkListFlags/pagination-limit-500-2  	     802	   7504184 ns/op	  859892 B/op	   35098 allocs/op
BenchmarkListFlags/pagination-2            	     190	  31245636 ns/op	  860305 B/op	   35108 allocs/op
BenchmarkListSegments/no-pagination-2      	      44	 132385529 ns/op	16565384 B/op	  725297 allocs/op
BenchmarkListSegments/pagination-limit-10-2         	     319	  18577732 ns/op	   29003 B/op	    1064 allocs/op
BenchmarkListSegments/pagination-limit-25-2         	     322	  18671660 ns/op	   53517 B/op	    2153 allocs/op
BenchmarkListSegments/pagination-limit-100-2        	     306	  19551521 ns/op	  176936 B/op	    7593 allocs/op
BenchmarkListSegments/pagination-limit-500-2        	     246	  24178912 ns/op	  835893 B/op	   36601 allocs/op
BenchmarkListSegments/pagination-2                  	     134	  44454415 ns/op	  836265 B/op	   36611 allocs/op
```

If i'm reading it correctly, it looks like this approach results in anywhere from a 2-10x speed improvement AND uses anywhere from half to a tenth of the memory allocations!!

And also it fixes the SQLite contention issue, as we are now setting the `MaxOpenConns(1)` if running with SQLite.

Seems like a win/win..win.

File: internal/storage/sql/db.go [hunk 2]
```diff
@@ -34,6 +34,13 @@ func Open(cfg config.Config, opts ...Option) (*sql.DB, Driver, error) {
 	if cfg.Database.MaxOpenConn > 0 {
 		sql.SetMaxOpenConns(cfg.Database.MaxOpenConn)
 	}
+
+	// if we're using sqlite, we need to set always set the max open connections to 1
+	// see: https://github.com/mattn/go-sqlite3/issues/274
+	if driver == SQLite {
+		sql.SetMaxOpenConns(1)
+	}
```
Comment (GeorgeMac) [inline internal/storage/sql/db.go:42]: How about we also log a warning suggesting we're going to ignore any attempts to manually configure this because of the driver choice:

```suggestion
	var maxOpenConn int 
	if cfg.Database.MaxOpenConn > 0 {
		maxOpenConn = cfg.Database.MaxOpenConn
	}

	// if we're using sqlite, we need to set always set the max open connections to 1
	// see: https://github.com/mattn/go-sqlite3/issues/274
	if driver == SQLite {
		if maxOpenConn > 1 {
			log.Warning("Ignoring config.db.max_open_conn due to driver limitation (sqlite)", zap.String("attempted_max_conn", maxOpenConn))
		}
		
		maxOpenConn = 1
	}
	
	sql.SetMaxOpenConns(maxOpenConn)
```

**Review by GeorgeMac — APPROVED** [review]
This is awesome. There is one suggested quality of life change to config.

**Comment by markphelps** [comment]
> This is awesome. There is one suggested quality of life change to config.

I moved the log warning to `internal/cmd/grpc` since we already have access to the logger there
